Make container removal fail if platform-specific containers fail

wk8 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

https://github.com/kubernetes/kubernetes/pull/74737 introduced a new in-memory
map for the dockershim, that could potentially (in pathological cases) cause
memory leaks - for containers that use GMSA cred specs, get created
successfully, but then never get started nor removed.

This patch addresses this issue by making container removal fail altogether
when platform-specific clean ups fail: this allows clean ups to be retried
later, when the kubelet attempts to remove the container again.

**Which issue(s) this PR fixes**:
Fixes #74843

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

wk8 commented:
/retrier ignore

wk8 commented:
/test pull-kubernetes-bazel-test

wk8 commented:
/test pull-kubernetes-bazel-build

wk8 commented:
/test pull-kubernetes-e2e-gce

wk8 commented:
/test pull-kubernetes-e2e-gce-100-performance

wk8 commented:
/test pull-kubernetes-kubemark-e2e-gce-big

wk8 commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

wk8 commented:
/test pull-kubernetes-e2e-gce-100-performance

cblecker commented:
@wk8 Why are the previous comments being deleted? Are you still scripting retests against kubernetes/kubernetes?

wk8 commented:
/test pull-kubernetes-bazel-test

wk8 commented:
@cblecker : deactivated the automated process. Sorry.

mattjmcnaughton reviewed (commented):
Thanks for your pr :)

Before I start a review, can you share more about this line in your pr: 

> that could potentially (in pathological cases) cause memory leaks - for containers that use GMSA cred specs, get created successfully, but then never get started nor removed.

Can you share more about these pathological cases and the size/scope of the memory leaks?

wk8 commented:
@mattjmcnaughton : the only way cred specs can be leak is if a container is created with a cred spec, then somehow never gets either started nor removed - and even then a single `containerCreationCleanupInfo` struct is leaked, which is really just a string. So for that leak to become significant, you'd need to have a large number of containers using cred specs that somehow get stuck forever in between creation and start time, all over a single Kubelet lifespan.

That's not very likely to occur, but it was deemed a sufficient concern to require some kind of GC before GMSA support is considered mature enough to be graduated to beta (see https://github.com/kubernetes/kubernetes/pull/74737#discussion_r261784181).

Please also note that this whole thing should become moot and superfluous when we can deprecate support for versions of docker older than the current latest release, which includes https://github.com/moby/moby/pull/38777, a patch that makes it possible to pass cred specs directly to docker instead of having to write somewhere, and then clean up.

mattjmcnaughton commented:
Gotcha, thanks @wk8 - just wanted to confirm that the failure was bad/common enough to warrant a fix and it sounds like we've already had that discussion.

cc @lavalamp how much context do you have on this part of the code base? I have fairly little. Happy to try and wade in if necessary, but just wanted to check with you first :)

PatrickLang commented on pkg/kubelet/dockershim/docker_container_cleanup_manager.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+// retryFailedCleanups retries running clean ups that have failed before.
+func (cm *containerCleanupManager) retryFailedCleanups() []string {
```
nit - it might be good to prefix this with unsafe so its consistent with the other functions that assume the caller holds the mutex

wk8 replied:
Good idea, done, thanks :)

yujuhong commented:
> That's not very likely to occur, but it was deemed a sufficient concern to require some kind of GC before GMSA support is considered mature enough to be graduated to beta (see [#74737 (comment)](https://github.com/kubernetes/kubernetes/pull/74737#discussion_r261784181)).
> 
> Please also note that this whole thing should become moot and superfluous when we can deprecate support for versions of docker older than the current latest release, which includes [moby/moby#38777](https://github.com/moby/moby/pull/38777), a patch that makes it possible to pass cred specs directly to docker instead of having to write somewhere, and then clean up.

@wk8 thanks for the PR. For the reasons you stated above, I think a simpler cleanup routine would suffice.

We don't need to worry about containers that were never started or removed -- kubelet should try to remove them, and would trigger the cleanup anyway. The reviewer comments from the previous PR was mainly about the case where the cleanup failed (for whatever reason), dockershim will never attempt to clean up again. To handle the specific corner, dockershim can periodically look at the cleanup that failed previously and attempt again. This should be a much smaller code change.


/cc @Random-Liu @liyanhui1228

wk8 commented:
> dockershim will never attempt to clean up again. To handle the specific corner, dockershim can periodically look at the cleanup that failed previously and attempt again. This should be a much smaller code change.

@yujuhong : that's indeed what this change is trying to accomplish, in a way that also stays reasonably efficient if for whatever reason we end up having a large number of such failures. Could you please be a little more specific as to what specific parts should be simplified? Thank you!

wk8 commented:
/test pull-kubernetes-integration

wk8 commented:
/test pull-kubernetes-integration

wk8 commented:
@yujuhong : simplified the implementation at the cost of a slightly decreased performance, please let me know what you think, thanks!

tedyu commented on pkg/kubelet/dockershim/docker_container_cleanup_manager.go:
```diff
@@ -0,0 +1,178 @@
+	cleanupInfo *containerCleanupInfo
+	// when this cleanup info was added to our state
+	timestamp int64
+	// how many times we've failed to actually run the cleanup associated
```
cleanup associated -> associated cleanup

wk8 replied:
Thanks, fixed :)

yujuhong commented:
/assign @Random-Liu

Random-Liu commented:
> The reviewer comments from the previous PR was mainly about the case where the cleanup failed (for whatever reason), dockershim will never attempt to clean up again. To handle the specific corner, dockershim can periodically look at the cleanup that failed previously and attempt again. This should be a much smaller code change.

@wk8 For this, I think you can simply let `performPlatformSpecificContainerForContainer` return error  in `RemoveContainer`, so that Kubelet will retry.

wk8 commented:
> @wk8 For this, I think you can simply let `performPlatformSpecificContainerForContainer` return error in `RemoveContainer`, so that Kubelet will retry.

@Random-Liu : clean ups really are a best-effort kind of thing, we don't necessarily want to fail container removal altogether if the clean up fails; besides, we want to clean up after containers after they were started, too, especially if they're long lived - and retry clean ups if they fail there too. In that case, we can't rely on `RemoveContainer` to do retries for us.

Since @yujuhong made the comment you quoted about this PR being too complicated, it's been simplified and @yujuhong has confirmed that she thinks the current shape is reasonable.

wk8 commented:
/test pull-kubernetes-integration

tedyu commented on pkg/kubelet/dockershim/docker_container_cleanup_manager.go:
```diff
@@ -0,0 +1,178 @@
+			containerIDsToCleanup = append(containerIDsToCleanup, containerID)
+
+			if withMetadata.timestamp <= timestampCutoff {
+				klog.Warningf("performing stale clean up for container %q", containerID)
```
I think this doesn't have to be warning - since we are cleaning up stale container

wk8 replied:
It shouldn't happen. Normally clean ups should be triggered after the container is successfully started, or removed. Stale clean ups are a sign that something is wrong.

tedyu commented on pkg/kubelet/dockershim/docker_container_cleanup_manager.go:
```diff
@@ -0,0 +1,178 @@
+)
+
+// If failing to clean up a container that many times, we'll just log an error and forget about this container.
+const maxFailures = 5
```
It seems the max failure count can be higher, say 9

Random-Liu commented:
> clean ups really are a best-effort kind of thing, we don't necessarily want to fail container removal altogether if the clean up fails;

The definition of "removal" is that all resources associated to the container should be gone. If this resource can't be released, remove should fail.

> besides, we want to clean up after containers after they were started, too, especially if they're long lived - and retry clean ups if they fail there too

@wk8 Can you explain what you are trying to cleanup here? Is it the in-memory entry? Or any files on disk?

I think as long as **everything is cleaned up after container is removed**, it is fine to have 1) an in-memory entry, or 2) a file on the disk for the container. Based on the density of containers on a node today, I don't think it will cause a big problem, at least doesn't worth the complexity and potential instability added in this PR.

Even though we want to add background cleanup, I hope it is tiny background goroutine and windows specific, because linux doesn't have this problem.

wk8 commented:
Thanks for your detailed answer @Random-Liu .

> The definition of "removal" is that all resources associated to the container is gone after that. If this resource can't be released, remove should fail by definition.

Even if what fails to be removed is tiny, doesn't actually hold any resources that might prevent anything else from happening in the future, and really just leaks tiny bits of non-sensitive information? The whole idea of these clean ups was from the beginning to be a best effort to get rid of non-sensitive, non-critical behind-the-scene stuff that is best removed, but really doesn't matter all that much if it's not.

As a k8s user, I wouldn't be too happy about container removals failing for things that don't matter to me in any way.

> And I believe that is also what @derekwaynecarr 's concern mentioned in [#74737 (comment)](https://github.com/kubernetes/kubernetes/pull/74737#discussion_r261428763).

I had understood @derekwaynecarr 's concern to be more about the number of pending cleanups that we keep track of in memory to grow large; and the rest of the discussion seems to focus more on that, too (see https://github.com/kubernetes/kubernetes/pull/74737#discussion_r261784181).

> @wk8 Can you explain what you are trying to cleanup here? Is it the in-memory entry? Or any files on disk?

As of now, clean ups are only used to get rid of temporary registry keys holding GMSA credentials (despite the name, those do not actually hold any sensitive data).

> Even if we want to add the background cleanup, I hope it is a tiny windows specific background goroutine, because linux doesn't have this problem.

Yes, it is Windows-specific, and tiny.

All that being said:
1. I'm not opposed to changing this patch to just run clean ups at removal time and simply fail removal if the clean up fails; provided that @yujuhong agrees with that.
2. This whole thing is a bit of a temporary fix around a limitation of older versions of Docker/containerd, that only accepts GMSA cred specs to be passed through files or registry keys. It can - and will - be removed when k8s drops support for these older Docker versions.

wk8 commented:
After discussing it with @Random-Liu on Slack, it seems just running clean ups at container removals, and failing those if the clean up fails, is the cleaner way to go here :)

Will make that change as soon as @yujuhong okays it.

yujuhong commented:
> After discussing it with @Random-Liu on Slack, it seems just running clean ups at container removals, and failing those if the clean up fails, is the cleaner way to go here

SGTM. Thanks!

wk8 committed:
Make container removal fail if platform-specific containers fail

https://github.com/kubernetes/kubernetes/pull/74737 introduced a new in-memory
map for the dockershim, that could potentially (in pathological cases) cause
memory leaks - for containers that use GMSA cred specs, get created
successfully, but then never get started nor removed.

This patch addresses this issue by making container removal fail altogether
when platform-specific clean ups fail: this allows clean ups to be retried
later, when the kubelet attempts to remove the container again.

Resolves issue https://github.com/kubernetes/kubernetes/issues/74843.

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@yujuhong & @Random-Liu : amended as requested to fail container removal when the clean up fails. PTAL, thanks!

Random-Liu commented on pkg/kubelet/dockershim/docker_container.go:
```diff
@@ -454,19 +455,27 @@ func (ds *dockerService) UpdateContainerResources(_ context.Context, r *runtimea
 	if cleanupInfo, present := ds.containerCleanupInfos[containerID]; present {
-		ds.performPlatformSpecificContainerCleanupAndLogErrors(containerID, cleanupInfo)
-		delete(ds.containerCleanupInfos, containerID)
+		errors = ds.performPlatformSpecificContainerCleanupAndLogErrors(containerID, cleanupInfo)
```
Shouldn't you call performPlatformSpecificContainerCleanup instead?

Or do you intentionally want to log the error?

Random-Liu reviewed (commented):
LGTM with one question

PatrickLang commented:
/priority important-soon
/sig windows
/milestone v1.16
